build.js

```diff
@@ -0,0 +1,126 @@
+const fs = require('fs');
+const path = require('path');
+/**
+ * The file extension of test files.
+ * @type {string}
+ */
+const fileExtension = '.json';
+/**
+ * The number of spaces to use for indentation in the output.
+ * @type {number}
+ */
+const jsonIndent = 2;
+/**
+ * The separator to use when combining test names.
+ * @type {string}
+ */
+const testNameSeparator = ', ';
+
+build();
+
+/**
+ * Build a combined test suite from a directory of test files.
+ * The output is written to stdout.
+ */
+function build() {
+    if (process.argv.length < 3) {
+        console.error('Usage: node build.js <test-folder-path>');
+        process.exit(1);
+    }
+
+    const testsFolder = process.argv[2];
+    const tests = readTestsFromDir(testsFolder);
+    const cts = {'tests': tests};
+    console.log(JSON.stringify(cts, null, jsonIndent));
+    console.error(`Wrote ${tests.length} tests to stdout.`);
+}
+
+
+/**
+ * Read all test files from a directory and its subdirectories.
+ * The directory name is prepended to the test name.
+ * @param dir {string} - The directory to read.
+ * @param relativePath {string[]} - The path to the directory.
+ * @returns {*[]} - An array of test objects.
+ */
+function readTestsFromDir(dir, relativePath = []) {
+    const files = fs.readdirSync(dir);
+    files.sort(filesBeforeDirs);
+    return files.reduce(addTestsFromFile, []);
+
+    /**
+     * Add tests from a file to the list of all tests.
+     * @param allTests - The list of all tests.
+     * @param file - The file to read.
+     * @returns {*} - The updated list of all tests.
+     */
+    function addTestsFromFile(allTests, file) {
+        const fullPath = path.join(dir, file);
+
+        if (isDir(file)) {
+            console.error(`Processing dir ${fullPath}`);
+            const tests = readTestsFromDir(fullPath, [...relativePath, file]);
+            return allTests.concat(tests);
+        }
+
+        if (path.extname(file) === fileExtension) {
+            console.error(`Processing file ${fullPath}`);
+            const basename = path.basename(file, fileExtension);
+            const testNamePrefix = [...relativePath, basename]
+                .join(testNameSeparator)
+                .replace('_', ' ');
+            const tests = readTestsFromFile(fullPath).map(prependToTestName(testNamePrefix));
+            return allTests.concat(tests);
+        }
+
+        return allTests;
+    }
+
+    /**
+     * Sort entries so that directories are listed after files.
```

Comment:
Would it be true to say that this function also sorts the files and the directories into some sort of lexicographic order? (Not sure what localeCompare does. In particular, I'm not sure if this would sort differently depending on the locale - probably not desirable. It would be good to have a sort that is stable across all environments.)

Reply:
Yes, that's how I understand it. A hard-coded locale can be provided to this method so it won't depend on the environment. I will make a update soon.

Reply:
Thanks. It's kind of strange to bring locale into the code at all, isn't it? Maybe node forces that.